Allow a button to trigger measurements in devices without sonar

0x2b3bfa0 opened the pull request:
**Connected to pin 7**

0x2b3bfa0 committed:
Modularize trigger function

0x2b3bfa0 commented on firmware/temperature_detection/user_config.h:
```diff
@@ -1,12 +1,16 @@
 #define USER_CONFIG_H
 
+// Switches
+#define ENABLE_SONAR
```
@1technophile, if you prefer to choose another name or wrap the switch with some code to detect M5StickC, feel free to push a new commit to this pull request.

1technophile replied:
I like it this way, we will use a `build_flag `on PIO to override the instruction on `user_config.h`

0x2b3bfa0 replied:
:shipit: then?

1technophile commented on firmware/temperature_detection/temperature_detection.ino:
```diff
@@ -61,29 +65,38 @@ void loop() {
-    } else if (distance <= DETECTION_DISTANCE) {
-        Serial.println("User is at a good distance");
-        detected_event = STATUS_DISTANCE_RIGHT;
+    if(digitalRead(BUTTON_PIN) == HIGH) {
+        detected_event = STATUS_TRIGGER;
```
By putting it there the button will have its status checked every TIME_BETWEEN_READINGS, requiring a long press for being triggered.
Maybe we should put the button check outside this function?

0x2b3bfa0 replied:
Yes, that's definitely not a feature; only happens when using the button in a sonar build.

0x2b3bfa0 replied:
~Moving the button conditional block just above the `ifdef` should fix it~

1technophile replied:
ok I will try that

0x2b3bfa0 replied:
I'm reluctant to move trigger logic outside the `trigger` function, but if the code grows more it might make sense. As you will.

1technophile replied:
As a first approach  for M5stickC we could also use TIME_BETWEEN_READINGS with a 0 value

0x2b3bfa0 replied:
Update: my proposed solution won't work.
5:00 AM, I'll propose a proper fix tomorrow.
To the prohibition of writing code late in the night I'll add one for the code reviews

1technophile replied:
Yes I agree, good night :-)

1technophile commented:
I will push the modification if you want

On Sun, Apr 19, 2020 at 8:53 PM Helio Machado <notifications@github.com>
wrote:

> *@0x2b3bfa0* commented on this pull request.
> ------------------------------
>
> In firmware/temperature_detection/user_config.h
> <https://github.com/helpfulengineering/project-temperature-detection/pull/13#discussion_r411044909>
> :
>
> > @@ -1,12 +1,16 @@
>  #ifndef USER_CONFIG_H
>  #define USER_CONFIG_H
>
> +// Switches
> +#define ENABLE_SONAR
>
> [image: :shipit:] then?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/helpfulengineering/project-temperature-detection/pull/13#discussion_r411044909>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADAV5XBOPJHBMAMV6KMPIFDRNOTIHANCNFSM4ML6QZFQ>
> .
>


-- 
Florian ROBERT
Phone: +1 251 895 6966

0x2b3bfa0 commented:
:+1:

El lun., 20 abr. 2020 4:08, Florian <notifications@github.com> escribió:

> I will push the modification if you want
>
> On Sun, Apr 19, 2020 at 8:53 PM Helio Machado <notifications@github.com>
> wrote:
>
> > *@0x2b3bfa0* commented on this pull request.
> > ------------------------------
> >
> > In firmware/temperature_detection/user_config.h
> > <
> https://github.com/helpfulengineering/project-temperature-detection/pull/13#discussion_r411044909
> >
> > :
> >
> > > @@ -1,12 +1,16 @@
> > #ifndef USER_CONFIG_H
> > #define USER_CONFIG_H
> >
> > +// Switches
> > +#define ENABLE_SONAR
> >
> > [image: :shipit:] then?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/helpfulengineering/project-temperature-detection/pull/13#discussion_r411044909
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ADAV5XBOPJHBMAMV6KMPIFDRNOTIHANCNFSM4ML6QZFQ
> >
> > .
> >
>
>
> --
> Florian ROBERT
> Phone: +1 251 895 6966
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/helpfulengineering/project-temperature-detection/pull/13#issuecomment-616270643>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACW4FW7BSNWHVDUCA3COGILRNOVB7ANCNFSM4ML6QZFQ>
> .
>

1technophile reviewed (commented):
Regarding the build flags I will push the modification in `platformio.ini` tomorrow

0x2b3bfa0 committed:
Fix modular trigger logic

0x2b3bfa0 committed:
Merge branch 'master' into trigger-button

1technophile committed:
Complete M5StickC board specific parameters

Currently the red led of the M5StickC goes ON when we press the button, the temperature measurement is took but the red led doesn't light ON a second time to indicate fever (TO DO)

1technophile committed:
Add sonar with Arduino boards on PIO

1technophile committed:
Add display time, trace and change on indicators

Add display time, if no delay the led doesn't have time to be seen
Add default ON status for button -M5stickC button is LOW when pushed)
Add trace for fever detected or not
Change limit fever to 30 so as to be able to test with the skin temperature
Change enum of fever status
Change display status checks

1technophile commented:
@0x2b3bfa0 Could you take a look on my last commits and give me some feedback please?

Especialy regarding indicators enum and checks done in `display_status `function. I'm not sure if I understood well the logic before.

Tests OK on UNO and M5StickC.

0x2b3bfa0 commented:
Sorry, @1technophile, I've done so many context switches in the last hours. I'm taking a look now 👀

1technophile commented:
No problem, there is no urgency

0x2b3bfa0 committed:
Fix logic

0x2b3bfa0 committed:
Fix platform quirks

0x2b3bfa0 commented:
@1technophile, I've made some changes and fixed some bitwise operator issues. Please check the code now and ask/revert what you find wrong.

1technophile commented:
Hi @0x2b3bfa0,

I'm having a build error with ESP 32 (M5StickC) environment regarding bitwise operation below:
```
D:/Users/Florian/Documents/GitHub/project-temperature-detection-orginal/firmware/temperature_detection/temperature_detection.ino: In function 'void display_status(status)':
D:/Users/Florian/Documents/GitHub/project-temperature-detection-orginal/firmware/temperature_detection/temperature_detection.ino:151:20: error: invalid conversion from 'int' to 'status' [-fpermissive]
         indicators = ~indicators;
                    ^
*** [.pio\build\m5stickc\src\temperature_detection.ino.cpp.o] Error 1
```

No problem on UNO or mega, do you have a proposal to avoid this error on ESP32 env?

0x2b3bfa0 committed:
Fix the negation type cast

0x2b3bfa0 commented:
@1technophile, fixed! I didn't C++ ;-)

1technophile committed:
remove INVERT_BUTTON, Stab time of sensor to 0 and ENABLE SONAR macro change

INVERT_BUTTON is finaly not necessary
-UENABLE_SONAR don't work on my env :-1:  so add -DENABLE_SONAR to other envs
Put SENSOR_STABILIZATION_TIME to 0 with M5StickC

1technophile commented:
OK now with the 3 envs thx!I I had to remove the INVERT_BUTTON mechanism, it is not necessary if we let it to LOW per default.
The stabilisation time should be at 0 with M5StickC so as to listen continuously the button state.
And unfortunately -UENABLE_SONAR doesn't work...

0x2b3bfa0 commented:
> OK now with the 3 envs thx!I I had to remove the INVERT_BUTTON mechanism, it is not necessary if we let it to LOW per default.

✅ Yes, I forgot we're forcing `INPUT_PULLUP`, so it would be nonsensical.

> The stabilisation time should be at 0 with M5StickC so as to listen continuously the button state.

❔ `SENSOR_STABILIZATION_TIME` does not affect the timing of the button/status polling loop, its only mission is waiting for the sensor to stabilize *once* the user presses the button or is at the right distance. I think that, if we disable it when triggering through a button, the results might vary.

> And unfortunately -UENABLE_SONAR doesn't work...

What a pity.

1technophile commented:
>❔ SENSOR_STABILIZATION_TIME does not affect the timing of the button/status polling loop, its only mission is waiting for the sensor to stabilize once the user presses the button or is at the right distance. I think that, if we disable it when triggering through a button, the results might vary.

Yes you are right, I propose just to decrease the value to 100ms for the moment, because at the current value it is too much time to wait before having the temperature measurement.
And regarding the sensor datasheet (p10) I don't think we need 3.5s.

1technophile committed:
Remove stabilization time macro and set it to 100ms

0x2b3bfa0 commented:
💯; :shipit:?

1technophile commented:
Are you OK with that. In my point of view we can squash the PR?

0x2b3bfa0 commented:
Ok, squash it. :-)
